service/interpreter/stateExecutionCounter.go

```diff
@@ -69,40 +69,94 @@ func (e *StateExecutionCounter) CreateNextExecutionId(stateId string) string {
 }
 
 func (e *StateExecutionCounter) MarkStateIdExecutingIfNotYet(stateReqs []StateRequest) error {
+	config := e.configer.Get()
+
 	needsUpdateSA := false
 	numOfNew := 0
 	for _, sr := range stateReqs {
 		if sr.IsResumeRequest() {
 			continue
 		}
 		s := sr.GetStateStartRequest()
-		numOfNew++
-		e.stateIdCurrentlyExecutingCounts[s.StateId]++
-		if e.stateIdCurrentlyExecutingCounts[s.StateId] == 1 {
-			// first time the stateId show up
-			needsUpdateSA = true
+
+		if e.globalVersioner.IsAfterVersionOfExecutingStateIdMode() {
+			switch mode := config.GetExecutingStateIdMode(); mode {
+			case "DISABLED":
+				// do nothing
+			case "ENABLED_FOR_ALL":
+				if e.increaseStateIdCurrentlyExecutingCounts(s) {
+					needsUpdateSA = true
+				}
+			default: // "ENABLED_FOR_STATES_WITH_WAIT_UNTIL" or nil or unrecognized enum value
+				options := s.GetStateOptions()
+				if !options.GetSkipWaitUntil() {
+					if e.increaseStateIdCurrentlyExecutingCounts(s) {
+						needsUpdateSA = true
+					}
+				}
+			}
+		} else {
+			if !config.GetDisableSystemSearchAttribute() {
+				if e.increaseStateIdCurrentlyExecutingCounts(s) {
+					needsUpdateSA = true
+				}
+			}
 		}
+
+		numOfNew++
 	}
 	e.totalCurrentlyExecutingCount += numOfNew
+
 	if needsUpdateSA {
 		return e.updateStateIdSearchAttribute()
 	}
 	return nil
 }
 
+func (e *StateExecutionCounter) increaseStateIdCurrentlyExecutingCounts(s iwfidl.StateMovement) bool {
+	e.stateIdCurrentlyExecutingCounts[s.StateId]++
+	// first time the stateId show up
+	return e.stateIdCurrentlyExecutingCounts[s.StateId] == 1
+}
+
 func (e *StateExecutionCounter) MarkStateExecutionCompleted(state iwfidl.StateMovement) error {
-	e.stateIdCurrentlyExecutingCounts[state.StateId]--
 	e.totalCurrentlyExecutingCount--
 
 	options := state.GetStateOptions()
 	skipStart := compatibility.GetSkipStartApi(&options)
 	e.continueAsNewCounter.IncExecutedStateExecution(skipStart)
 
+	config := e.configer.Get()
+
+	if e.globalVersioner.IsAfterVersionOfExecutingStateIdMode() {
+		switch mode := config.GetExecutingStateIdMode(); mode {
+		case "DISABLED":
+			return nil
+		case "ENABLED_FOR_ALL":
+			e.decreaseStateIdCurrentlyExecutingCounts(state)
+		default: // "ENABLED_FOR_STATES_WITH_WAIT_UNTIL" or nil or unrecognized enum value
+			if options.GetSkipWaitUntil() {
```

Comment:
We should use `skipStart := compatibility.GetSkipStartApi(&options)` like above for compatibity. Or just use the above resolved value.

Reply:
Done